advanced_source/cpp_extension.rst

```diff
@@ -0,0 +1,1033 @@
+    auto d_gates = at::zeros_like(gate_weights);
+
+    const auto batch_size = new_cell.size(0);
+    const auto state_size = new_cell.size(1);
+
+    const int threads = 1024;
+    const dim3 blocks((state_size + threads - 1) / threads, batch_size);
+
+    AT_DISPATCH_FLOATING_TYPES(X.type(), "lltm_forward_cuda", ([&] {
+      lltm_cuda_backward_kernel<scalar_t><<<blocks, threads>>>(
+          d_old_cell.data<scalar_t>(),
+          d_gates.data<scalar_t>(),
+          grad_h.contiguous().data<scalar_t>(),
+          grad_cell.contiguous().data<scalar_t>(),
+          new_cell.contiguous().data<scalar_t>(),
+          input_gate.contiguous().data<scalar_t>(),
+          output_gate.contiguous().data<scalar_t>(),
+          candidate_cell.contiguous().data<scalar_t>(),
+          gate_weights.contiguous().data<scalar_t>(),
+          state_size);
+    }));
+
+    auto d_weights = d_gates.t().mm(X);
+    auto d_bias = d_gates.sum(/*dim=*/0, /*keepdim=*/true);
+
+    auto d_X = d_gates.mm(weights);
+    auto d_old_h = d_X.slice(/*dim=*/1, 0, state_size);
+    auto d_input = d_X.slice(/*dim=*/1, state_size);
+
+    return {d_old_h, d_input, d_weights, d_bias, d_old_cell, d_gates};
+  }
+
+Integrating a C++/CUDA Operation with PyTorch
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Integration of our CUDA-enabled op with PyTorch is again very straightforward.
+If you want to write a ``setup.py`` script, it could look like this::
+
+  from setuptools import setup
+  from torch.utils.cpp_extension import BuildExtension, CUDAExtension
+
+  setup(
+      name='lltm',
+      ext_modules=[
+          CUDAExtension('lltm_cuda', [
+              'lltm_cuda.cpp',
+              'lltm_cuda_kernel.cu',
+          ])
+      ],
+      cmdclass={
+          'build_ext': BuildExtension
+      })
+
+Instead of :func:`CppExtension`, we now use :func:`CUDAExtension`. We can just
+specify the ``.cu`` file along with the ``.cpp`` files -- the library takes
+care of all the hassle this entails for you. The JIT mechanism is even
+simpler::
+
+  from torch.utils.cpp_extension import load
+
+  lltm = load(name='lltm', sources=['lltm_cuda.cpp', 'lltm_cuda_kernel.cu'])
+
+Performance Comparison
+**********************
+
+Our hope was that parallelizing and fusing the pointwise operations of our code
+with CUDA would improve the performance of our LLTM. Let's see if that holds
+true. We can run the code I listed earlier to run a benchmark. Our fastest
+version earlier was the CUDA-based C++ code::
+
+  Forward: 149.802 us | Backward 393.458 us
+
+
+And now with our custom CUDA kernel::
+
+  Forward: 129.431 us | Backward 304.641 us
+
+More performance increases!
+
+Conclusion
+----------
+
+You should now be equipped with a good overview of PyTorch's C++ extension
+mechanism as well as a motivation for using them. You can find the code examples
+displayed in this note `here
+<https://github.com/goldsbborough/pytorch-cpp-extension>`_. If you have
```

Comment:
Is this intentionally not the link to https://github.com/pytorch/extension-cpp/ ?

Reply:
Nope, this needs to be fixed.